docs/config.md
<!--
Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# Configuring Model Navigator

The Model Navigator can be configured with a [YAML](https://yaml.org/) file or via the command-line interface (CLI).
Every flag supported by the CLI is supported in the configuration file, but some configurations are only supported using
the config file.

The placeholders below are used throughout the configuration:

* `<boolean>`: a boolean that can take `true` or `false` as value.
* `<string>`: a regular string
* `<comma-delimited-list>`: a list of comma separated items.
* `<int>`: a regular integer value.
* `<list>`: a list of values.

## CLI and YAML Config Options

A list of all the configuration options supported by both the CLI and YAML config file are shown below. Brackets
indicate that a parameter is optional. For non-list and non-object parameters the value is set to the specified default.

The CLI flags corresponding to each of the options below are obtained by converting the `snake_case` options
to `--kebab-case`. For example, `model_name` in the YAML would be `--model-name` in the CLI.

```yaml
- Name of model
model_name: <string>

- Path to the model file/directory
model_path: <string>

# The directory where Model Navigator stores the artifacts and temporary files
[ workspace_path: <string> | default: 'workspace' ]

# Number of top final configurations selected from analysis
[ top_n_configs: <int> | default: 3 ]

# Max concurrency used for config search in analysis
[ max_concurrency: <int> | default: 1024 ]

# Max number of model instances used for config search in analysis
[ max_instance_count: <int> | default: 5 ]

# Max preferred batch size used for config search in analysis
[ max_preferred_batch_size: <int> | default: 32 ]

# Concurrency values to be used for the analysis
[ concurrency: <comma-delimited-string|list> | default: None ]

# Number of model instances values to be used for the analysis
[ instance_counts: <comma-delimited-string|list> | default: None ]

# Max latency constraint in milliseconds
[ max_latency_ms: <int> | default: None ]

# Min throughput constraint in inference/seconds
[ min_throughput: <int> | default: None ]

# Max used GPU memory constraint in MB
[ max_gpu_usage_mb: <int> | default: None ]

# The amount of workspace the `ICudaEngine` uses in the optimization
[ max_workspace_size: <int> | default: None ]

# Configure TensorRT builder for precision layer selection
[ target_precisions: <comma-delimited-string|list> | default: "fp16,tf32" ]

# Generate ONNX graphs that use only ops available in a given opset
[ onnx_opsets: <comma-delimited-string|list> | default: "12,13" ]

# The minimum shapes TensorRT optimization profile(s) supports
[ min_shapes: <list> | default: None ]

# The optimal shapes TensorRT optimization profile(s) supports
[ opt_shapes: <list> | default: None ]

# The maximum shapes TensorRT optimization profile(s) supports. Also, defines shapes of input data used during performance analysis.
[ max_shapes: <list> | default: None ]

# Range of values used during performance analysis defined per input. Also, defines shapes of input data used during performance analysis.
[ value_ranges: <list> | default: None ]

# The model inputs and their shapes
[ inputs: <list> | default: None ]

# The model outputs and their shapes
[ outputs: <list> | default: None ]

# Relative tolerance parameter for output comparison
[ rtol: <list> | default: None ]

# Absolute tolerance parameter for output comparison
[ atol: <list> | default: None ]
-
# How Model Navigator will launch Triton Inference Server. It should be either "docker" or "local".
[ triton_launch_mode: <string> ]

# The full path to the `tritonserver` binary executable
[ triton_server_path: <string> | default: tritonserver ]

# The version of Triton Inference Server container (only `triton_launch_mode: docker`)
[ triton_version: <string> | default: 20.12-py3 ]

# The protocol used to communicate with Triton Inference Server. Only 'http' and 'grpc' are allowed for the values.
[ client_protocol: <string> | default: grpc ]

# Logging level
[ verbose: <bool> | default: false ]

# List of GPU UUIDs to be used for the profiling. Use 'all' to profile all the GPUs visible by CUDA.
[ gpus: <string|comma-delimited-list-string> | default: 'all' ]

# Specify path to config yaml file
[ config_file: <string> ]
```

## YAML Only Options

The following config options are supported only by the YAML config file.

```yaml

# List of preferred batch sizes used in analysis - list of values
[ preferred_batch_sizes: <list<list>> | default: None ]

# List of objectives that the user wants to sort the results by
[ objectives: <objective|list> ]

```

## Example usage

Using CLI:

```shell
$ model-navigator --model-name add_sub \
   --model-path model-navigator/examples/quick-start/model.pt \
   --inputs INPUT__0:-1,16:float32 INPUT__1:-1,16:float32 \
   --outputs OUTPUT__0:-1,16:float32 OUTPUT__1:-1,16:float32 \
   --max-concurrency 256 \
   --max-latency-ms 50 \
   --verbose
```

Using YAML:

```yaml
model_name: add_sub
model_path: model-navigator/examples/quick-start/model.pt
inputs:
 - INPUT__0:16:float32
 - INPUT__1:16:float32
outputs:
 - OUTPUT__0:16:float32
 - OUTPUT__1:16:float32
max_concurrency: 256
max_latency_ms: 50
verbose: true
```

Running Model Navigator:

```
$ model-navigator -f config.yaml
```

## Objectives and Preferred Batch Sizes

Manual settings preferred batch sizes and objectives:

```yaml
objectives:
   - perf_througput
   - perf_latency
prefferred_batch_sizes:
   - [ 16, 32 ]
   - [ 32, 64 ]
```

Weighted objectives:

```yaml
objectives:
   perf_througput: 5
   perf_latency: 10
```


model_navigator/config/config.py
# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from typing import Any, List, Optional

import abc
from enum import Enum

import yaml

from ..model_navigator_exceptions import ModelNavigatorException
from ..record.record import RecordType
from .config_field import ConfigField
from .config_list_generic import ConfigListGeneric
from .config_list_numeric import ConfigListNumeric
from .config_list_string import ConfigListString
from .config_object import ConfigObject
from .config_primitive import ConfigPrimitive
from .config_union import ConfigUnion


def _objective_list_output_mapper(objectives):
    # Takes a list of objectives and maps them
    # into a dict
    output_dict = {}
    for objective in objectives:
        value = ConfigPrimitive(type_=int)
        value.set_value(10)
        output_dict[objective] = value
    return output_dict


class ModelNavigatorBaseConfig(metaclass=abc.ABCMeta):
    def __init__(self, *, required_keys: Optional[List[str]] = None):
        """
        Create a new config.
        """

        self._fields = {}
        self._required_keys = required_keys or []
        self._fill_config()

    def __getattr__(self, name):
        try:
            return self._fields[name].value()
        except KeyError:
            raise AttributeError(f"Missing {name} attribute")

    def __setattr__(self, name: str, value: Any) -> None:
        if "_fields" in vars(self) and name in self._fields:
            if isinstance(value, List) and isinstance(value[0], Enum):
                value = [v.value for v in value]
            elif isinstance(value, Enum):
                value = value.value
            self._fields[name].set_value(value)
        else:
            super().__setattr__(name, value)

    def __getstate__(self):
        return self._fields

    def __setstate__(self, state):
        self._fields = state

    def set_config_values(self, args):
        """
        Set the config values. This function sets all the values for the
        config. CLI arguments have the highest priority, then YAML config
        values and then default values.

        Parameters
        ----------
        args : argparse.Namespace
            Parsed arguments from the CLI

        Raises
        ------
        ModelNavigatorException
            If the required fields are not specified, it will raise
            this exception
        """

        # Config file has been specified
        if "config_file" in args:
            yaml_config = self._load_config_file(args.config_file)
        else:
            yaml_config = None
        for key, value in self._fields.items():
            self._fields[key].set_name(key)
            if key in args:
                self._fields[key].set_value(getattr(args, key))
            elif yaml_config is not None and key in yaml_config:
                self._fields[key].set_value(yaml_config[key])
            elif value.default_value() is not None:
                self._fields[key].set_value(value.default_value())
            elif value.required():
                flags = ", ".join(value.flags())
                raise ModelNavigatorException(
                    f"Config for {value.name()} is not specified. You need to specify it using the YAML config file or using the {flags} flags in CLI."
                )

    def get_config(self):
        """
        Get the config dictionary.

        Returns
        -------
        dict
            Returns a dictionary where the keys are the
            configuration name and the values are ConfigField objects.
        """

        return self._fields

    def get_all_config(self):
        """
        Get a dictionary containing all the configurations.

        Returns
        -------
        dict
            A dictionary containing all the configurations.
        """

        config_dict = {}
        for config in self._fields.values():
            config_dict[config.name()] = config.value()

        return config_dict

    def _add_config(self, config_field):
        """
        Add a new config field.

        Parameters
        ----------
        config_field : ConfigField
            Config field to be added

        Raises
        ------
        KeyError
            If the field already exists, it will raise this exception.
        """

        if config_field.name() in self._required_keys:
            config_field.set_required(True)

        if config_field.name() not in self._fields:
            self._fields[config_field.name()] = config_field
        else:
            raise KeyError

    def _load_config_file(self, file_path):
        """
        Load YAML config

        Parameters
        ----------
        file_path : str
            Path to the Model Navigator config file
        """

        with open(file_path, "r") as config_file:
            config = yaml.load(config_file, Loader=yaml.FullLoader)
            return config

    def merge(self, other_config: "ModelNavigatorBaseConfig"):
        conflicting_fields = {
            name: field
            for name, field in other_config.get_config().items()
            if name in self.get_config() and self.get_config()[name] != other_config.get_config()[name]
        }
        if conflicting_fields:
            raise KeyError(f"Conflicting fields found during configuration merge: {', '.join(conflicting_fields)}")

        self._fields.update(**other_config._fields)

        return self

    @abc.abstractmethod
    def _fill_config(self):
        pass

    def __str__(self) -> str:
        return ", ".join([f"{key}={field.value()}" for key, field in self._fields.items()])


class ModelNavigatorConfig(ModelNavigatorBaseConfig):
    """
    Model Navigator config object.
    """

    def _fill_config(self):
        # common config fields
        self._add_config(
            ConfigField(
                "model_name",
                flags=["--model-name"],
                field_type=ConfigPrimitive(str, required=True),
                description="Name of model.",
            )
        )

        self._add_config(
            ConfigField(
                "model_path",
                flags=["--model-path"],
                default_value="",
                field_type=ConfigPrimitive(str, required=True),
                description="Path to file with model.",
            )
        )
        self._add_config(
            ConfigField(
                "config_file",
                field_type=ConfigPrimitive(str),
                flags=["-f", "--config-file"],
                description="Path to Model Navigator Config File.",
            )
        )
        self._add_config(
            ConfigField(
                "workspace_path",
                flags=["--workspace-path"],
                default_value="workspace",
                field_type=ConfigPrimitive(str),
                description="Path to output directory.",
            )
        )
        self._add_config(
            ConfigField(
                "verbose",
                field_type=ConfigPrimitive(bool),
                parser_args={"action": "store_true"},
                default_value=False,
                flags=["--verbose"],
                description="Enable verbose mode.",
            )
        )

        # analyzer specific fields
        self._add_config(
            ConfigField(
                "top_n_configs",
                field_type=ConfigPrimitive(int),
                flags=["--top-n-configs"],
                description="Number of top final configurations selected from analysis.",
                default_value=3,
            )
        )

        self._add_config(
            ConfigField(
                "max_concurrency",
                flags=["--max-concurrency"],
                field_type=ConfigPrimitive(int),
                default_value=1024,
                description="Max concurrency used for config search in analysis",
            )
        )
        self._add_config(
            ConfigField(
                "max_instance_count",
                flags=["--max-instance-count"],
                field_type=ConfigPrimitive(int),
                default_value=5,
                description="Max number of model instances used for config search in analysis",
            )
        )
        self._add_config(
            ConfigField(
                "max_preferred_batch_size",
                flags=["--max-preferred-batch-size"],
                field_type=ConfigPrimitive(int),
                default_value=32,
                description="Max preferred batch size used for config search in analysis",
            )
        )

        # User defined list of parameters
        self._add_config(
            ConfigField(
                "concurrency",
                flags=["--concurrency"],
                field_type=ConfigListNumeric(int),
                description="Concurrency values to be used for the analysis in manual mode",
            )
        )

        self._add_config(
            ConfigField(
                "instance_counts",
                flags=["--instance-counts"],
                field_type=ConfigListNumeric(int),
                description="Comma-delimited list of instance counts to use for the profiling.",
            )
        )
        self._add_config(
            ConfigField(
                "preferred_batch_sizes",
                flags=["--preferred-batch-sizes"],
                field_type=ConfigListGeneric(ConfigListNumeric(int)),
                description="List of batch sizes to use for the profiling in dynamic batching.",
            )
        )

        self._add_config(
            ConfigField(
                "max_latency_ms",
                flags=["--max-latency-ms"],
                field_type=ConfigPrimitive(int),
                description="Maximal latency in ms that analyzed model should match.",
            )
        )

        self._add_config(
            ConfigField(
                "min_throughput",
                flags=["--min-throughput"],
                field_type=ConfigPrimitive(int),
                description="Minimal throughput that analyzed model should match.",
            )
        )
        self._add_config(
            ConfigField(
                "max_gpu_usage_mb",
                flags=["--max-gpu-usage-mb"],
                field_type=ConfigPrimitive(int),
                description="Maximal GPU memory usage in MB that analyzed model should match.",
            )
        )

        objectives_scheme = ConfigUnion(
            [
                ConfigObject(
                    schema={tag: ConfigPrimitive(type_=int) for tag in RecordType.get_all_record_types().keys()}
                ),
                ConfigListString(output_mapper=_objective_list_output_mapper),
            ]
        )

        self._add_config(
            ConfigField(
                "objectives",
                field_type=objectives_scheme,
                default_value={"perf_throughput": 10},
                description="Model Navigator uses the objectives described here to find the best configuration for model.",
            )
        )

        self._add_config(
            ConfigField(
                "triton_version",
                flags=["--triton-version"],
                field_type=ConfigPrimitive(str),
                default_value="21.03-py3",
                description="Triton Server Docker version",
            )
        )
        self._add_config(
            ConfigField(
                "triton_launch_mode",
                field_type=ConfigPrimitive(str),
                flags=["--triton-launch-mode"],
                default_value="local",
                choices=["local", "docker"],
                description="The method by which to launch Triton Server. "
                "'local' assumes tritonserver binary is available locally. "
                "'docker' pulls and launches a triton docker container with "
                "the specified version.",
            )
        )
        self._add_config(
            ConfigField(
                "triton_server_path",
                field_type=ConfigPrimitive(str),
                flags=["--triton-server-path"],
                default_value="tritonserver",
                description="The full path to the tritonserver binary executable",
            )
        )

        self._add_config(
            ConfigField(
                "client_protocol",
                flags=["--client-protocol"],
                choices=["http", "grpc"],
                field_type=ConfigPrimitive(str),
                default_value="grpc",
                description="The protocol used to communicate with the Triton Inference Server",
            )
        )
        self._add_config(
            ConfigField(
                "gpus",
                flags=["--gpus"],
                field_type=ConfigListString(),
                default_value="all",
                description="List of GPU UUIDs to be used for the profiling. "
                "Use 'all' to profile all the GPUs visible by CUDA.",
            )
        )
